app/javascript/src/js/formValidations.js

```diff
@@ -1,121 +1,39 @@
-      if (formErrors) {
-        ev.preventDefault();
-        handleFormSubmit(form);
-        ev.stopPropagation();
-      }
-    });
-
-    const inputs = form.querySelectorAll('.form__element');
-
-    inputs.forEach((item) => {
-      item.addEventListener('change', function (ev) {
-        const errors = validate(form, constraints, { fullMessages: false }) || {};
-        showErrorsForInput(this, errors[this.name]);
-      });
-    });
-  });
-});
-
-function handleFormSubmit(form) {
-  const formErrors = validate(form, constraints, { fullMessages: false });
-  if (!formErrors) {
-    form.submit();
+      message: "^is too short (minimum is 4 characters)"
+    }},
+    password: { presence: true, length: {minimum: 6, maximum: 128,  message: "^is too short (minimum is 6 characters)" } }
   }
-  showErrors(form, formErrors || {});
-}
 
-function showErrors(form, errors) {
-  const formInputs = form.querySelectorAll('.form__element');
-  formInputs.forEach((item) => {
-    showErrorsForInput(item, errors[item.name]);
-  });
-}
+  static validators = { passwordIsConfirmed: { attributes: ['password_confirmation' ]  } }
 
-function showErrorsForInput(input, errors) {
-  const inputParent = input.parentNode;
-  const errorWrapper = findErrorWrapper(input) || createErrorWrapper(input, inputParent);
-  const errorDiv = findErrorDiv(input) || createErrorDiv(input, inputParent);
-
-  if (!errors) {
-    errorWrapper.parentNode.insertBefore(input, errorWrapper);
-    errorWrapper.parentNode.removeChild(errorWrapper);
-    errorDiv.parentNode.removeChild(errorDiv);
-    return;
+  passwordIsConfirmed({attr, value}) {
+    if (value !== document.getElementById('user_password').value) {
+	    this.errors.add(attr, 'The passwords do not match')
+    }
   }
 
-  errorWrapper.classList.add('field_with_errors');
-  errorDiv.innerHTML = errors;
-}
-
-function findErrorWrapper(input) {
-  if (input.parentNode.classList.contains('field_with_errors')) {
-    return input.parentNode;
+  afterValidate({ el, attr }) {
+    this.errorMessageEl(el).textContent = this.errorMessage(attr)
   }
-  return null;
-}
 
-function createErrorWrapper(input, inputParent) {
-  const newDiv = document.createElement('div');
-  inputParent.appendChild(newDiv);
-  newDiv.appendChild(input);
-  return newDiv;
-}
+  errorMessageEl(el) {
+    return el.closest(".form-field").querySelector(".form__error-message")
+  }
 
-function findErrorDiv(input) {
-  return document.querySelector(`.${input.id}`);
+  errorMessage(attr) {
+    return this.errors.has(attr) ? this.errors.get(attr)[0] : ""
+  }
 }
 
-function createErrorDiv(input, inputParent) {
-  const divName = document.createElement('div');
-  divName.classList.add('form__error-message', 'push-down', input.id);
-  inputParent.parentNode.insertBefore(divName, inputParent.nextSibling);
-  return divName;
-}
+const application = Application.start();
+application.register('signup-validation', SignupValidation)
```

Comment:
Stimulus controllers conventionally live in the `app/javascript/controllers` directory. Can we move this file to that directory and rename it to `form_validation_controller.js` please?

Then we should be able to remove these lines for registering the controller. That will be automatically done for us in the `controller/index.js` file.

Reply:
this no longer works when removing these 2 lines. it does bring it in as shown here

![image](https://user-images.githubusercontent.com/24649589/184511124-4b1d8ba9-dc2c-4025-83e7-b1ea0f7457b9.png)

![image](https://user-images.githubusercontent.com/24649589/184511154-b9f435d6-7537-42e3-ad83-18a4ab24126e.png)

but the validation no longer works. something to do with the stimulus-validation package maybe?

Reply:
Yeah I just noticed the library is a little out of date. It looks like it was built for Stimulus 2.0 but we're on 3. My fault, for not looking into that library enough before suggesting it  😅

But it's working and is so much better and cleaner than what we had before. Let's get this in and revisit somewhere down the line.